python/across_api/burstcube/schema.py

```diff
@@ -0,0 +1,219 @@
+# Copyright © 2023 United States Government as represented by the
+# Administrator of the National Aeronautics and Space Administration.
+# All Rights Reserved.
+
+
+import hashlib
+import json
+from decimal import Decimal
+from enum import Enum
+from typing import List, Optional
+
+import astropy.units as u  # type: ignore
+from astropy.time import Time  # type: ignore
+from pydantic import ConfigDict, computed_field, model_validator  # type: ignore
+
+from ..base.schema import (
+    AstropySeconds,
+    AstropyTime,
+    BaseSchema,
+    JsonStr,
+    OptionalDateRangeSchema,
+    OptionalPositionSchema,
+)
+
+
+class TOOReason(str, Enum):
+    """
+    Reasons for rejecting TOO observations
+
+    Attributes
+    ----------
+    saa
+        In SAA
+    earth_occult
+        Earth occulted
+    moon_occult
+        Moon occulted
+    sun_occult
+        Sun occulted
+    too_old
+        Too old
+    other
+        Other
+    none
+        None
+    """
+
+    saa = "In SAA"
+    earth_occult = "Earth occulted"
+    moon_occult = "Moon occulted"
+    sun_occult = "Sun occulted"
+    too_old = "Too old"
+    other = "Other"
+    none = "None"
+
+
+class TOOStatus(str, Enum):
+    """
+    Enumeration class representing the status of a Target of Opportunity (TOO) request.
+
+    Attributes:
+    requested
+        The TOO request has been submitted.
+    rejected
+        The TOO request has been rejected.
+    declined
+        The TOO request has been declined.
+    approved
+        The TOO request has been approved.
+    executed
+        The TOO request has been executed.
+    other
+        The TOO request has a status other than the predefined ones.
+    """
+
+    requested = "Requested"
+    rejected = "Rejected"
+    declined = "Declined"
+    approved = "Approved"
+    executed = "Executed"
+    deleted = "Deleted"
+    other = "Other"
+
+
+class BurstCubeTriggerInfo(BaseSchema):
+    """
+    Metadata schema for the BurstCube Target of Opportunity (TOO) request. Note
+    that this schema is not strictly defined, keys are only suggested, and
+    additional keys can be added as needed.
+    """
+
+    trigger_name: Optional[str] = None
+    trigger_mission: Optional[str] = None
+    trigger_instrument: Optional[str] = None
+    trigger_id: Optional[str] = None
+    trigger_duration: Optional[AstropySeconds] = None
+    classification: Optional[str] = None
+    justification: Optional[str] = None
+
+    model_config = ConfigDict(extra="allow")
+
+    @model_validator(mode="before")
+    def convert_json_string_to_dict(cls, data):
+        if isinstance(data, str):
+            return json.loads(data)
+        return data
```

Comment:
Why are we converting it to and from a string? Why not just store it as a dictionary?

Reply:
It needs to be converted to a JSON string so it can be written to DynamoDB under the key `trigger_info`. When it gets read back it gets converted back into a dict.

Reply:
Ah, I see the confusion. DynamoDB records are JSON documents. It does not need to be converted to a string.

Reply:
So I figured out that the reason I was doing this was due to a problem with how I was writing TOOs to DynamoDB. The main issue being the need to have numbers be `Decimal`. I had been handling this with a conversion Schema called `BurstCubeTOOModel`, but this broke when adding `BurstCubeTriggerInfo` as a sub class, and my solution was to convert it to a JSON string. This worked, but as you point out, wasn't necessary.

So I got rid of `BurstCubeTOOModel` and now just convert `BurstCubeTOOSchema` into a dict with all `float`s converted to `Decimal`. This works, allowed me to delete `BurstCubeTOOModel` and a bunch of other code, and is much cleaner. 

Fixed by [2c9c6f3](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/2c9c6f3059423867cdfecb464ae0cdc06a766622) and [b02735d](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/b02735d5c2ae03ff2381fc41dfb1c8b67f323e9d). 

I need to push an update to the tests as they are now failing.

Reply:
So I had to put this code back in. [ce858aa](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1975/commits/ce858aabffed8445410dfc1091a41e82fd05e4f1)

I had forgotten part of why this exists. It's due to the complexities of HTTP POST. POST does not allow you to submit files and JSON data, so you have to submit any JSON data as a file, which means it's encoded as a JSON `str`, not as a `dict`. This means that it fails `model_validate` unless you put in a validator (like the removed code) that parses JSON `str` into a `dict`. 

This is either a limitation of FastAPI or there is some esoteric way around it I cannot figure out. I found one discussion that suggests the former: https://github.com/tiangolo/fastapi/discussions/9868

This highlights the need for tests which submit through the API running in a sandbox environment, rather than just testing at the python class level, because a change that breaks the API interface right now is not going to be caught.